esmtools/timeutils.py

```diff
@@ -0,0 +1,251 @@
+        elif self.type == "CFTimeIndex":
+            calendar = get_calendar(self._obj)
+            x = cftime.date2num(self._obj, "days since 1990-01-01", calendar=calendar)
+            x = xr.DataArray(x, dims=self._obj.dims, coords=self._obj.coords)
+            return x
+        else:
+            raise ValueError("DataArray is not a time array of datetimes or cftimes.")
+
+    @staticmethod
+    def construct_quarterly_aliases():
+        quarters = ["Q", "BQ", "QS", "BQS"]
+        for month in [
+            "JAN",
+            "FEB",
+            "MAR",
+            "APR",
+            "MAY",
+            "JUN",
+            "JUL",
+            "AUG",
+            "SEP",
+            "OCT",
+            "NOV",
+            "DEC",
+        ]:
+            quarters.append(f"Q-{month}")
+            quarters.append(f"BQ-{month}")
+            quarters.append(f"BQS-{month}")
+            quarters.append(f"QS-{month}")
+        return quarters
+
+    @staticmethod
+    def construct_annual_aliases():
+        years = ["A", "Y", "BA", "BY", "AS", "YS", "BAS", "BYS", "Q"]
+        for month in [
+            "JAN",
+            "FEB",
+            "MAR",
+            "APR",
+            "MAY",
+            "JUN",
+            "JUL",
+            "AUG",
+            "SEP",
+            "OCT",
+            "NOV",
+            "DEC",
+        ]:
+            years.append(f"A-{month}")
+            years.append(f"BA-{month}")
+            years.append(f"BAS-{month}")
+            years.append(f"AS-{month}")
+        return years
+
+    def construct_slope_factors(self):
+        years = self.construct_annual_aliases()
+        years = {k: self.annual_factor for k in years}
+
+        quarters = self.construct_quarterly_aliases()
+        quarters = {k: self.annual_factor / 4 for k in quarters}
+
+        months = ("M", "BM", "CBM", "MS", "BMS", "CBMS")
+        months = {k: self.annual_factor / 12 for k in months}
+
+        semimonths = {k: 15 for k in ("SM", "SMS")}
+
+        weeks = ("W", "W-SUN", "W-MON", "W-TUE", "W-WED", "W-THU", "W-FRI", "W-SAT")
+        weeks = {k: 7 for k in weeks}
+
+        days = {k: 1 for k in ("B", "C", "D")}
+        hours = {k: 1 / 24 for k in ("BH", "H")}
+        mins = {k: 1 / (24 * 60) for k in ("T", "min")}
+        secs = {k: 1 / (24 * 60 * 60) for k in ("S")}
+        millisecs = {k: 1 / (24 * 60 * 60 * 1e3) for k in ("ms", "L")}
+        microsecs = {k: 1 / (24 * 60 * 60 * 1e6) for k in ("U", "us")}
+        nanosecs = {k: 1 / (24 * 60 * 60 * 1e9) for k in ("N")}
+
+        DATETIME_FACTOR = {}
+        for d in (
+            years,
+            quarters,
+            months,
+            semimonths,
+            weeks,
+            days,
+            hours,
+            mins,
+            secs,
+            millisecs,
+            microsecs,
+            nanosecs,
+        ):
+            DATETIME_FACTOR.update(d)
+        return DATETIME_FACTOR
+
+
+def get_calendar(ds, dim=None):
```

Comment:
I started an xarray PR on this. 

Easier:
from .times import infer_calendar_name
infer_calendar_name(self)

Reply:
They don't import that globally so I can't do `from xarray.times import ...`. But that `infer_calendar_name` function is basically identical to what I do. But I didn't realize I could just do `.calendar`. I'll just replicate it here.